### PR TITLE
Dispose

### PR DESCRIPTION
**Pull request by ChristianTrolleMikkelsen**
Ensure dispose is called and refactored.
Also inheritance is now so that dispose is accessible on IConsumer/IProducer.

### CODE REVIEW

File: src/Franz/HighLevel.fs [hunk 1]
```diff
@@ -249,14 +253,18 @@ type ConsumerOffsetManagerV0(topicName, brokerRouter : BrokerRouter) =
         LogConfiguration.Logger.Info.Invoke(sprintf "Offsets fetched from Zookeeper: %A" offsets)
         offsets
 
+    let handleOffsetCommitResponseCodes (offsetCommitResponse : OffsetCommitResponse) (offsets : seq<PartitionOffset>) (managerName : string) =
+        let errorCodes = Seq.concat (offsetCommitResponse.Topics |> Seq.map (fun t -> t.Partitions |> Seq.filter (fun p -> p.ErrorCode <> ErrorCode.NoError) |> Seq.map(fun p -> sprintf "Topic: %s Partition: %i ErrorCode: %s" t.Name p.Id (p.ErrorCode.ToString()))))
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: We can only get one error code per request

File: src/Franz/HighLevel.fs [hunk 2]
```diff
@@ -289,6 +297,7 @@ module internal ErrorHelper =
 
 /// Offset manager for version 1. This commits and fetches offset to/from Kafka broker.
 type ConsumerOffsetManagerV1(topicName, brokerRouter : BrokerRouter) =
+
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Why this extra line?

File: src/Franz/HighLevel.fs [hunk 3]
```diff
@@ -6,10 +6,24 @@ open Franz
 open System.Collections.Concurrent
 open Franz.Compression
 
-exception ClusterErrorException of string * int
-exception RequestTimedOutException of string * int
-exception BrokerReturnedErrorException of string * ErrorCode
-exception ErrorCommitingOffsetException of string * OffsetCommitResponse
+type ErrorCommittingOffsetException (errorCodes : seq<string>) =
+    inherit Exception()
+    member e.Codes = errorCodes
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Only one error code is possible per request

File: src/Franz/HighLevel.fs [hunk 4]
```diff
@@ -562,6 +583,9 @@ type ConsumerOptions() =
 
 [<AbstractClass>]
 type BaseConsumer(topicName, brokerRouter : BrokerRouter, consumerOptions : ConsumerOptions) =
+
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Why this extra line?

File: src/Franz/HighLevel.fs (hunk 3 above)
Comment (ChristianTrolleMikkelsen) [inline src/Franz/HighLevel.fs]: Are you sure? The response contains partition information for all X number of partitions, so in theory, there could be multiple errors, one for each partition?

File: src/Franz/HighLevel.fs (hunk 2 above)
Comment (ChristianTrolleMikkelsen) [inline src/Franz/HighLevel.fs]: Fixed.

File: src/Franz/HighLevel.fs [hunk 5]
```diff
@@ -7,10 +7,13 @@ open System.Collections.Concurrent
 open Franz.Compression
 open Franz.Internal
 
-type ErrorCommittingOffsetException (errorCodes : seq<string>) =
+type ErrorCommittingOffsetException (offsetManagerName : string, topic : string, consumerGroup : string, errorCodes : seq<string>) =
     inherit Exception()
     member e.Codes = errorCodes
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Only one errorcode can happen at a time

File: src/Franz/HighLevel.fs [hunk 6]
```diff
@@ -5,11 +5,28 @@ open System.Collections.Generic
...
 open Franz.Compression
+open Franz.Internal
 
-exception ClusterErrorException of string * int
-exception RequestTimedOutException of string * int
-exception BrokerReturnedErrorException of string * ErrorCode
-exception ErrorCommitingOffsetException of string * OffsetCommitResponse
+type ErrorCommittingOffsetException (offsetManagerName : string, topic : string, consumerGroup : string, errorCodes : seq<string>) =
+    inherit Exception()
+    member e.Codes = errorCodes
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Only on error code can happen at a time

File: src/Franz/HighLevel.fs [hunk 7]
```diff
@@ -85,7 +103,7 @@ type Producer(brokerRouter : BrokerRouter, compressionCodec : CompressionCodec,
             innerSend key messages topicName requiredAcks brokerProcessingTimeout retryCount
         | ErrorCode.RequestTimedOut ->
             retryOnRequestTimedOut (fun x -> innerSend key messages topicName requiredAcks (brokerProcessingTimeout * 2) (retryCount + 1)) retryCount
-        | _ -> raise(BrokerReturnedErrorException("Received broker response with error code", partitionResponse.ErrorCode))
+        | _ -> raiseWithLog(BrokerReturnedErrorException partitionResponse.ErrorCode)
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Are we sure this is not log in different places?

File: src/Franz/HighLevel.fs [hunk 8]
```diff
@@ -249,14 +266,19 @@ type ConsumerOffsetManagerV0(topicName, brokerRouter : BrokerRouter) =
         LogConfiguration.Logger.Info.Invoke(sprintf "Offsets fetched from Zookeeper: %A" offsets)
         offsets
 
+    let handleOffsetCommitResponseCodes (offsetCommitResponse : OffsetCommitResponse) (offsets : seq<PartitionOffset>) (consumerGroup : string) (managerName : string) =
+        let errorCodes = Seq.concat (offsetCommitResponse.Topics |> Seq.map (fun t -> t.Partitions |> Seq.filter (fun p -> p.ErrorCode <> ErrorCode.NoError) |> Seq.map(fun p -> sprintf "Topic: %s Partition: %i ErrorCode: %s" t.Name p.Id (p.ErrorCode.ToString()))))
+        let topic = (offsetCommitResponse.Topics |> Seq.head).Name
+        match Seq.isEmpty errorCodes with
+        | false -> raiseWithLog(ErrorCommittingOffsetException(managerName, topic, consumerGroup, errorCodes))
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Are we sure this is not log in different places?

File: src/Franz/HighLevel.fs [hunk 9]
```diff
@@ -332,6 +353,13 @@ type ConsumerOffsetManagerV1(topicName, brokerRouter : BrokerRouter) =
             LogConfiguration.Logger.Info.Invoke(sprintf "Offsets fetched from Kafka (using Kafka): %A" offsets)
             offsets
 
+    let handleOffsetCommitResponseCodes (offsetCommitResponse : OffsetCommitResponse) (offsets : seq<PartitionOffset>) (consumerGroup : string) (managerName : string) =
+        let errorCodes = Seq.concat (offsetCommitResponse.Topics |> Seq.map (fun t -> t.Partitions |> Seq.filter (fun p -> p.ErrorCode <> ErrorCode.NoError) |> Seq.map(fun p -> sprintf "Topic: %s Partition: %i ErrorCode: %s" t.Name p.Id (p.ErrorCode.ToString()))))
+        let topic = (offsetCommitResponse.Topics |> Seq.head).Name
+        match Seq.isEmpty errorCodes with
+        | false -> raiseWithLog(ErrorCommittingOffsetException(managerName, topic, consumerGroup, errorCodes))
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Are we sure this is not log in different places?

File: src/Franz/HighLevel.fs [hunk 10]
```diff
@@ -421,6 +445,13 @@ type ConsumerOffsetManagerV2(topicName, brokerRouter : BrokerRouter) =
             LogConfiguration.Logger.Info.Invoke(sprintf "Offsets fetched from Kafka (using KafkaV2): %A" offsets)
             offsets
 
+    let handleOffsetCommitResponseCodes (offsetCommitResponse : OffsetCommitResponse) (offsets : seq<PartitionOffset>) (consumerGroup : string) (managerName : string) =
+        let errorCodes = Seq.concat (offsetCommitResponse.Topics |> Seq.map (fun t -> t.Partitions |> Seq.filter (fun p -> p.ErrorCode <> ErrorCode.NoError) |> Seq.map(fun p -> sprintf "Topic: %s Partition: %i ErrorCode: %s" t.Name p.Id (p.ErrorCode.ToString()))))
+        let topic = (offsetCommitResponse.Topics |> Seq.head).Name
+        match Seq.isEmpty errorCodes with
+        | false -> raiseWithLog(ErrorCommittingOffsetException(managerName, topic, consumerGroup, errorCodes))
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Are we sure this is not log in different places?

File: src/Franz/HighLevel.fs [hunk 11]
```diff
@@ -661,7 +689,7 @@ type BaseConsumer(topicName, brokerRouter : BrokerRouter, consumerOptions : Cons
                     partitionOffsets.AddOrUpdate(partitionId, new Func<Id, Offset>(fun _ -> earliestOffset), fun _ _ -> earliestOffset) |> ignore
                     return! self.ConsumeInChunks(partitionId, maxBytes)
                 | _ ->
-                    raise(BrokerReturnedErrorException("Received broker response with error code", partitionResponse.ErrorCode))
+                    raiseWithLog(BrokerReturnedErrorException partitionResponse.ErrorCode)
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Are we sure this is not log in different places?

File: src/Franz/HighLevel.fs (hunk 4 above)
Comment (ChristianTrolleMikkelsen) [inline src/Franz/HighLevel.fs]: Fixed

File: src/Franz/HighLevel.fs [hunk 12]
```diff
@@ -103,7 +103,7 @@ type Producer(brokerRouter : BrokerRouter, compressionCodec : CompressionCodec,
             innerSend key messages topicName requiredAcks brokerProcessingTimeout retryCount
         | ErrorCode.RequestTimedOut ->
             retryOnRequestTimedOut (fun x -> innerSend key messages topicName requiredAcks (brokerProcessingTimeout * 2) (retryCount + 1)) retryCount
-        | _ -> raiseWithLog(BrokerReturnedErrorException partitionResponse.ErrorCode)
+        | _ -> raiseWithFatalLog(BrokerReturnedErrorException partitionResponse.ErrorCode)
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Should this not be an error? Franz works after this exception

File: src/Franz/HighLevel.fs [hunk 13]
```diff
@@ -85,7 +85,7 @@ type Producer(brokerRouter : BrokerRouter, compressionCodec : CompressionCodec,
 
     let retryOnRequestTimedOut retrySendFunction (retryCount : int) =
         LogConfiguration.Logger.Warning.Invoke(sprintf "Producer received RequestTimedOut on Ack from Brokers, retrying (%i) with increased timeout" retryCount, RequestTimedOutException())
-        if retryCount > 1 then raise(RequestTimedOutRetryExceededException())
+        if retryCount > 1 then raiseWithFatalLog(RequestTimedOutRetryExceededException())
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Should this not be an error? Franz works after this exception

File: src/Franz/HighLevel.fs [hunk 14]
```diff
@@ -63,7 +63,7 @@ type BaseProducer private(brokerRouter : BrokerRouter, topicName, compressionCod
         | ErrorCode.NotLeaderForPartition ->
             brokerRouter.RefreshMetadata()
             send key messages requiredAcks brokerProcessingTimeout
-        | _ -> raiseWithLog(BrokerReturnedErrorException partitionResponse.ErrorCode)
+        | _ -> raiseWithFatalLog(BrokerReturnedErrorException partitionResponse.ErrorCode)
```
Comment (kichristensen) [inline src/Franz/HighLevel.fs]: Should this not be an error? Franz works after this exception

**Pull request merged**